File: apps/android/app/src/test/java/net/auralis/app/features/search/SearchFiltersTest.kt
package net.auralis.app.features.search

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Pure behaviour tests for [SearchFilterState]'s chip logic — the Kotlin mirror of
 * `apps/web/src/features/search/searchFilters.ts`. No Compose, no Android framework type
 * anywhere in this file or the class under test.
 */
class SearchFiltersTest {
    @Test
    fun `podcasts primary reveals no secondary row`() {
        assertTrue(secondaryFilterOptions("podcasts").isEmpty())
    }

    @Test
    fun `all primary reveals no secondary row`() {
        assertTrue(secondaryFilterOptions("all").isEmpty())
    }

    @Test
    fun `music primary reveals songs albums and artists`() {
        assertEquals(listOf("all", "songs", "albums", "artists"), secondaryFilterOptions("music").map { it.value })
    }

    @Test
    fun `books primary reveals books series and authors`() {
        assertEquals(listOf("all", "books", "series", "authors"), secondaryFilterOptions("books").map { it.value })
    }

    @Test
    fun `selecting a primary resets the secondary filter`() {
        // Music's "artists" is meaningless once Books is selected — carrying it across would
        // silently leave the new secondary row with a value none of its own chips represent,
        // and visibleKinds would then hide every book result. This is the regression this test
        // exists to pin: without the reset in selectPrimaryFilter, this assertion fails.
        val afterArtists = selectSecondaryFilter(selectPrimaryFilter(DEFAULT_SEARCH_FILTER_STATE, "music"), "artists")
        assertEquals(SearchFilterState("music", "artists"), afterArtists)

        val afterSwitchingToBooks = selectPrimaryFilter(afterArtists, "books")
        assertEquals(SearchFilterState("books", "all"), afterSwitchingToBooks)
    }

    @Test
    fun `selecting the already-active primary chip clears back to all`() {
        val selected = selectPrimaryFilter(DEFAULT_SEARCH_FILTER_STATE, "music")
        val cleared = selectPrimaryFilter(selected, "music")
        assertEquals(DEFAULT_SEARCH_FILTER_STATE, cleared)
    }

    @Test
    fun `selecting the already-active secondary chip clears it back to all`() {
        val withSecondary = selectSecondaryFilter(SearchFilterState("music", "all"), "songs")
        assertEquals("songs", withSecondary.secondary)
        val cleared = selectSecondaryFilter(withSecondary, "songs")
        assertEquals("all", cleared.secondary)
    }

    @Test
    fun `no selection shows every kind grouped by content type`() {
        val visible = visibleKinds("all", "all")
        assertTrue(visible.books)
        assertTrue(visible.podcasts)
        assertTrue(visible.artists)
        assertTrue(visible.albums)
        assertTrue(visible.tracks)
        // Series/authors only ever surface once Books is explicitly selected — the unfiltered
        // view groups at the same three content types it always has.
        assertTrue(!visible.series)
        assertTrue(!visible.authors)
    }

    @Test
    fun `books with no secondary shows books series and authors together`() {
        val visible = visibleKinds("books", "all")
        assertTrue(visible.books)
        assertTrue(visible.series)
        assertTrue(visible.authors)
        assertTrue(!visible.podcasts)
        assertTrue(!visible.artists)
    }

    @Test
    fun `books narrowed to series shows only series`() {
        val visible = visibleKinds("books", "series")
        assertTrue(visible.series)
        assertTrue(!visible.books)
        assertTrue(!visible.authors)
    }

    @Test
    fun `books narrowed to authors shows only authors`() {
        val visible = visibleKinds("books", "authors")
        assertTrue(visible.authors)
        assertTrue(!visible.books)
        assertTrue(!visible.series)
    }

    @Test
    fun `music with no secondary shows artists albums and tracks together`() {
        val visible = visibleKinds("music", "all")
        assertTrue(visible.artists)
        assertTrue(visible.albums)
        assertTrue(visible.tracks)
        assertTrue(!visible.books)
    }

    @Test
    fun `music narrowed to songs shows only tracks`() {
        val visible = visibleKinds("music", "songs")
        assertTrue(visible.tracks)
        assertTrue(!visible.artists)
        assertTrue(!visible.albums)
    }

    @Test
    fun `music narrowed to albums shows only albums`() {
        val visible = visibleKinds("music", "albums")
        assertTrue(visible.albums)
        assertTrue(!visible.tracks)
        assertTrue(!visible.artists)
    }

    @Test
    fun `music narrowed to artists shows only artists`() {
        val visible = visibleKinds("music", "artists")
        assertTrue(visible.artists)
        assertTrue(!visible.albums)
        assertTrue(!visible.tracks)
    }

    @Test
    fun `podcasts primary shows only podcasts regardless of secondary`() {
        val visible = visibleKinds("podcasts", "all")
        assertTrue(visible.podcasts)
        assertTrue(!visible.books)
        assertTrue(!visible.artists)
    }

    @Test
    fun `an unrecognised primary degrades to showing everything rather than throwing`() {
        val visible = visibleKinds("something-unknown", "all")
        assertTrue(visible.books)
        assertTrue(visible.podcasts)
        assertTrue(visible.artists)
    }

    @Test
    fun `an unrecognised secondary under books degrades to the books-with-no-secondary view`() {
        val visible = visibleKinds("books", "something-unknown")
        assertTrue(visible.books)
        assertTrue(visible.series)
        assertTrue(visible.authors)
    }
}

File: apps/web/src/features/search/searchFilters.test.ts
import { describe, expect, it } from 'vitest';
import {
  DEFAULT_SEARCH_FILTER_STATE,
  secondaryFilterOptions,
  selectPrimaryFilter,
  selectSecondaryFilter,
  visibleKinds,
} from './searchFilters.js';

describe('secondaryFilterOptions', () => {
  it('reveals All/Songs/Albums/Artists for Music', () => {
    expect(secondaryFilterOptions('music')).toEqual([
      { value: 'all', label: 'All' },
      { value: 'songs', label: 'Songs' },
      { value: 'albums', label: 'Albums' },
      { value: 'artists', label: 'Artists' },
    ]);
  });

  it('reveals All/Books/Series/Authors for Books', () => {
    expect(secondaryFilterOptions('books')).toEqual([
      { value: 'all', label: 'All' },
      { value: 'books', label: 'Books' },
      { value: 'series', label: 'Series' },
      { value: 'authors', label: 'Authors' },
    ]);
  });

  it('has no second row for Podcasts — the search API only ever returns whole shows', () => {
    expect(secondaryFilterOptions('podcasts')).toEqual([]);
  });

  it('has no second row when nothing is selected', () => {
    expect(secondaryFilterOptions('all')).toEqual([]);
  });

  it('degrades to no second row for an unrecognised primary rather than throwing', () => {
    expect(secondaryFilterOptions('bogus')).toEqual([]);
  });
});

describe('selectPrimaryFilter', () => {
  it('selects a new primary and resets the secondary to all', () => {
    const afterMusic = selectPrimaryFilter(DEFAULT_SEARCH_FILTER_STATE, 'music');
    expect(afterMusic).toEqual({ primary: 'music', secondary: 'all' });

    const midway = selectSecondaryFilter(afterMusic, 'albums');
    expect(midway).toEqual({ primary: 'music', secondary: 'albums' });

    // Switching primary to Books drops the Music-only "albums" secondary —
    // it isn't even a valid Books option.
    const afterBooks = selectPrimaryFilter(midway, 'books');
    expect(afterBooks).toEqual({ primary: 'books', secondary: 'all' });
  });

  it('clicking the already-active primary clears the row back to all, clearing the secondary too', () => {
    const musicWithSongs = selectSecondaryFilter(
      selectPrimaryFilter(DEFAULT_SEARCH_FILTER_STATE, 'music'),
      'songs',
    );
    expect(musicWithSongs).toEqual({ primary: 'music', secondary: 'songs' });

    const cleared = selectPrimaryFilter(musicWithSongs, 'music');
    expect(cleared).toEqual({ primary: 'all', secondary: 'all' });
    // The secondary row is gone entirely once primary is 'all' — nothing
    // left over from the "songs" pick.
    expect(secondaryFilterOptions(cleared.primary)).toEqual([]);
  });

  it('selecting All directly clears both rows regardless of prior state', () => {
    const booksWithSeries = selectSecondaryFilter(
      selectPrimaryFilter(DEFAULT_SEARCH_FILTER_STATE, 'books'),
      'series',
    );
    expect(selectPrimaryFilter(booksWithSeries, 'all')).toEqual({
      primary: 'all',
      secondary: 'all',
    });
  });
});

describe('selectSecondaryFilter', () => {
  it('selects a secondary value', () => {
    const state = selectPrimaryFilter(DEFAULT_SEARCH_FILTER_STATE, 'books');
    expect(selectSecondaryFilter(state, 'authors')).toEqual({
      primary: 'books',
      secondary: 'authors',
    });
  });

  it('clicking the already-active secondary clears it back to all', () => {
    const state = selectSecondaryFilter(
      selectPrimaryFilter(DEFAULT_SEARCH_FILTER_STATE, 'books'),
      'authors',
    );
    expect(selectSecondaryFilter(state, 'authors')).toEqual({ primary: 'books', secondary: 'all' });
  });
});

describe('visibleKinds', () => {
  it('shows books, podcasts and all three music kinds when nothing is selected, but not series/authors', () => {
    expect(visibleKinds('all', 'all')).toEqual({
      books: true,
      series: false,
      authors: false,
      podcasts: true,
      artists: true,
      albums: true,
      tracks: true,
    });
  });

  it('Books + All shows books, series and authors together', () => {
    expect(visibleKinds('books', 'all')).toEqual({
      books: true,
      series: true,
      authors: true,
      podcasts: false,
      artists: false,
      albums: false,
      tracks: false,
    });
  });

  it.each([
    ['books', { books: true, series: false, authors: false }],
    ['series', { books: false, series: true, authors: false }],
    ['authors', { books: false, series: false, authors: true }],
  ])('Books + %s narrows to just that kind', (secondary, expected) => {
    const result = visibleKinds('books', secondary);
    expect(result.books).toBe(expected.books);
    expect(result.series).toBe(expected.series);
    expect(result.authors).toBe(expected.authors);
    expect(result.podcasts).toBe(false);
    expect(result.artists).toBe(false);
    expect(result.albums).toBe(false);
    expect(result.tracks).toBe(false);
  });

  it('Music + All shows artists, albums and tracks together', () => {
    expect(visibleKinds('music', 'all')).toEqual({
      books: false,
      series: false,
      authors: false,
      podcasts: false,
      artists: true,
      albums: true,
      tracks: true,
    });
  });

  it.each([
    ['songs', { artists: false, albums: false, tracks: true }],
    ['albums', { artists: false, albums: true, tracks: false }],
    ['artists', { artists: true, albums: false, tracks: false }],
  ])('Music + %s narrows to just that kind', (secondary, expected) => {
    const result = visibleKinds('music', secondary);
    expect(result.artists).toBe(expected.artists);
    expect(result.albums).toBe(expected.albums);
    expect(result.tracks).toBe(expected.tracks);
    expect(result.books).toBe(false);
    expect(result.series).toBe(false);
    expect(result.authors).toBe(false);
    expect(result.podcasts).toBe(false);
  });

  it('Podcasts shows only podcasts, regardless of secondary (it has no second row)', () => {
    expect(visibleKinds('podcasts', 'all')).toEqual({
      books: false,
      series: false,
      authors: false,
      podcasts: true,
      artists: false,
      albums: false,
      tracks: false,
    });
  });

  it('degrades an unrecognised primary to the same "show everything" result as all, rather than throwing', () => {
    expect(() => visibleKinds('bogus', 'all')).not.toThrow();
    expect(visibleKinds('bogus', 'all')).toEqual(visibleKinds('all', 'all'));
  });
});
